Allow global install and use external modules

cebe opened the pull request:
currently it is not possible to install uipf on a system with `make install` and also other modules which are not directly located in uipf development directory can not be used.
This pull request adds some changes to improve the situation. It is work in progress and should not be merged.
- [x] Add install targets ot CMakeLists.txt
- [x] Change ModuleManager to be able to load modules from different locations
- [ ] open bug (opencv waitkey does not work in console module):
  
  ```
  uipf: symbol lookup error: /Code/build/libVisualizationModule.so: undefined symbol: _ZN2cv7waitKeyEi
  ```
- [x] fix chdir() issue for console too (to make relative paths in yaml files work)
- [x] Add Documentation on how to load modules
- [x] Add Documentation on how to compile external modules

cebe committed:
allow uipf to be installed globally

- added CMake target for installation
- adding module search paths

cebe committed:
make external modules work

cebe committed:
enhance datatype constructors

add default and copy constructors

cebe committed:
Update extending-modules.md

cebe committed:
fixed another "missing symbol" error

akarsh commented:
Is the issue OS environment specific or library dependency or file path miss match specific in a brief overview ?
I can test on Mac OS 
if you want to have overview.  
As, i installed opencv library with XCODE & I have a small test with cmakelists successful to check for blurrying an image through console arguments.

```
cmake_minimum_required(VERSION 2.8)
project( BlurImage )
find_package( OpenCV )
include_directories( ${OpenCV_INCLUDE_DIRS} )
add_executable( BlurImage BlurImage.cpp )
target_link_libraries( BlurImage ${OpenCV_LIBS} )
```

cebe commented:
these changes might fix the issue described in #114 so you might try to install on mac again. however the `make install` command may not install it to the correct directories as I do not know where to best store things on mac. But maybe CMake already handles this correctly...

cebe committed:
make types const compatible

cebe committed:
improved exception catching

cebe committed:
make Context work in external modules

cebe committed:
improved logging output

cebe committed:
Locale independent parameter parsing

atof() depends on the locale, which means, that on systems with e.g. de_DE,
parameters with dot as decimal symbol are not parsed correctly because atof() expects
a comma instead.
This fixes it to use dot always.

See also http://stackoverflow.com/questions/1333451/locale-independent-atof?lq=1

cebe committed:
added .desktop files

this allows opening yaml files from applications in gnome and KDE and also
opening them from the file browser.

cebe committed:
ensure console command can run on a server without display.

http://stackoverflow.com/questions/17979185/qt-5-1-qapplication-without-display-qxcbconnection-could-not-connect-to-displ

cebe committed:
GUI chdir to current yaml file

this is needed to make relative paths in the yaml file work.

cebe committed:
build with optimization

cebe committed:
exclude IDEA cmake build

cebe committed:
chdir to the yaml file when loading a config to make relative paths work

cebe committed:
docs about compiling a module
